src/consensus/tx_verify.cpp

```diff
@@ -272,24 +273,66 @@ inline bool checkOutput(const CTxOut &out, CValidationState &state, CAmount &nVa
             return state.DoS(100, false, REJECT_INVALID, "bad-assets-transfer-amount");
         }
 
-        std::pair <std::string, uint16_t> asset(assetTransfer.assetId, assetTransfer.uniqueId);
-        if (nAssetVin.count(asset))
-            nAssetVin[asset] += assetTransfer.nAmount;
+        if (nAssetVin.count(assetTransfer.assetId))
+            nAssetVin[assetTransfer.assetId] += assetTransfer.nAmount;
         else
-            nAssetVin.insert(std::make_pair(asset, assetTransfer.nAmount));
+            nAssetVin.insert(std::make_pair(assetTransfer.assetId, assetTransfer.nAmount));
 
-        //check unique asset amount, should be alway 1 COIN
-        if (assetTransfer.isUnique && nAssetVin[asset] != 1 * COIN)
-            return state.DoS(100, false, REJECT_INVALID, "bad-asset-inputvalues-outofrange");
+        if (assetTransfer.isUnique) {
+            uint64_t idRange = assetTransfer.uniqueId + assetTransfer.nAmount / COIN;
+
+            if (!nMapids.count(assetTransfer.assetId))
+                nMapids.insert({assetTransfer.assetId, {}});
+
+            nMapids[assetTransfer.assetId].emplace_back(std::make_pair(assetTransfer.uniqueId, idRange));
+        }
 
-        if (!MoneyRange(assetTransfer.nAmount) || !MoneyRange(nAssetVin.at(asset))) {
+        if (!MoneyRange(assetTransfer.nAmount) || !MoneyRange(nAssetVin.at(assetTransfer.assetId))) {
             return state.DoS(100, false, REJECT_INVALID, "bad-asset-inputvalues-outofrange");
         }
     }
 
     return true;
 }
 
+inline bool checkAssetsOutputs(CValidationState &state, std::map <std::string, CAmount> nAssetVin,
+                                    std::map <std::string, CAmount> nAssetVout,
+                                    std::map <std::string, std::vector<std::pair<uint64_t, uint64_t>>> nVinIds, 
+                                    std::map <std::string, std::vector<std::pair<uint64_t, uint64_t>>> nVoutIds) {
+    //check vin/vout amouts
+    for (const auto &outValue: nAssetVout) {
+        if (!nAssetVin.count(outValue.first) || nAssetVin.at(outValue.first) != outValue.second)
+            return state.DoS(100, false, REJECT_INVALID, "bad-asset-inputs-outputs-mismatch");
+    }
+    //Check for missing outputs
+    for (const auto &inValue: nAssetVin) {
+        if (!nAssetVout.count(inValue.first) || nAssetVout.at(inValue.first) != inValue.second)
+            return state.DoS(100, false, REJECT_INVALID, "bad-asset-inputs-outputs-mismatch");
+    }
+
+    for (auto &vin : nVinIds){
+        vin.second = combineUniqueIdPairs(vin.second);
+    }
+    
+    for (auto &vout : nVoutIds){
+        vout.second = combineUniqueIdPairs(vout.second);
+    }
+
+    //check uniqueids
+    for (auto vin : nVinIds){
```

Comment:
this loops is inefficient could cause stall node while performance this check for multiple tx with large unique ids. would be possible to hash ids in a way to prove it match?

Reply:
i think a unordered_map can be used here, let me test